### PR TITLE
Rework csv validation, warnings, headers etc

### PR DESCRIPTION
**Pull request by jamesjefferies**

### CODE REVIEW

File: app/controllers/admin/amr_uploaded_readings_controller.rb [hunk 1]
```diff
@@ -4,6 +4,9 @@ class AmrUploadedReadingsController < AdminController
 
     def show
       @amr_uploaded_reading = AmrUploadedReading.find(params[:id])
+
```
Comment (jamesjefferies) [inline app/controllers/admin/amr_uploaded_readings_controller.rb]: Break out valid data and data with warnings for user to peruse

File: app/services/amr/data_feed_validator.rb [hunk 2]
```diff
@@ -7,13 +7,21 @@ def initialize(config, array_of_rows)
...
-      @array_of_rows.shift(@config.number_of_header_rows)
+      handle_header
       @array_of_rows.delete_if { |row| invalid_row?(row) }
       @array_of_rows
     end
 
   private
 
+    def handle_header
+      if @array_of_rows.first.join(',') == @config.header_example
```
Comment (jamesjefferies) [inline app/services/amr/data_feed_validator.rb]: If first row is the header example, shift it out and carry on

File: app/services/amr/data_feed_validator.rb [hunk 3]
```diff
@@ -7,13 +7,21 @@ def initialize(config, array_of_rows)
...
+        @array_of_rows.shift
+      elsif @config.number_of_header_rows
```
Comment (jamesjefferies) [inline app/services/amr/data_feed_validator.rb]: We still need this for highlands which has multiple header rows

File: app/controllers/admin/amr_uploaded_readings_controller.rb [hunk 4]
```diff
@@ -26,12 +28,18 @@ def create
...
         @errors = @amr_reading_data.errors.messages[:reading_data].join(', ')
+        set_valid_readings_and_warnings
         render :new
       end
     end
 
     private
 
+    def set_valid_readings_and_warnings
+      @valid_reading_data = @amr_uploaded_reading.reading_data.reject { |reading| reading.key?('warning') || reading.key?(:warning) }
```
Comment (jamesalmond) [inline app/controllers/admin/amr_uploaded_readings_controller.rb]: Push this into the model? Indifferent access?

File: app/controllers/admin/amr_uploaded_readings_controller.rb [hunk 5]
```diff
@@ -26,12 +28,18 @@ def create
...
       else
         @errors = @amr_reading_data.errors.messages[:reading_data].join(', ')
+        set_valid_readings_and_warnings
         render :new
       end
     end
 
     private
 
+    def set_valid_readings_and_warnings
```
Comment (jamesalmond) [inline app/controllers/admin/amr_uploaded_readings_controller.rb]: Should these be limited here?

File: app/models/amr_data_feed_import_log.rb [hunk 6]
```diff
@@ -19,10 +19,14 @@
...
   belongs_to :amr_data_feed_config
 
-  scope :errored,    -> { where.not(error_messages: nil) }
-  scope :successful, -> { where(error_messages: nil) }
+  scope :errored,       -> { where.not(error_messages: nil) }
+  scope :successful,    -> { where(error_messages: nil) }
+  scope :with_warnings, -> { includes(:amr_reading_warnings).where.not(amr_reading_warnings: { id: nil }) }
 
   has_many :meters, -> { distinct }, through: :amr_data_feed_readings
+  has_many :amr_reading_warnings
```
Comment (jamesalmond) [inline app/models/amr_data_feed_import_log.rb]: Duplicated

File: app/models/amr_reading_data.rb [hunk 7]
```diff
@@ -3,84 +3,92 @@ class AmrReadingData
...
+      end
+    end
   end
 
-  def is_there_an_invalid_reading_date?
-    @reading_data.each { |reading| Date.strptime(reading[:reading_date], @date_format) }
-  rescue ArgumentError
-    lenient_date_checking
+  def missing_readings?(readings)
+    readings.compact.size < (48 - @missing_reading_threshold)
```
Comment (jamesalmond) [inline app/models/amr_reading_data.rb]: Does this catch empty strings?

File: app/models/amr_reading_data.rb [hunk 8]
```diff
@@ -3,84 +3,92 @@ class AmrReadingData
...
-  def all_the_reading_dates_are_dates?
-    @reading_data.reject { |reading| reading[:reading_date].instance_of? Date }.blank?
+  def invalid_row_check
+    @reading_data.each do |reading|
+      mpan_mprn = reading[:mpan_mprn]
+      reading_date = reading[:reading_date]
+      readings = reading[:readings]
+
+      if missing_readings?(readings)
+        reading[:warning] = :missing_readings
```
Comment (jamesalmond) [inline app/models/amr_reading_data.rb]: `reading[:warnings] = []`

File: app/models/amr_reading_warning.rb [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+#
+# Foreign Keys
+#
+#  fk_rails_...  (amr_data_feed_import_log_id => amr_data_feed_import_logs.id) ON DELETE => cascade
+#
+
+class AmrReadingWarning < ApplicationRecord
+  belongs_to :amr_data_feed_import_log
+
+  enum warning: [:blank_readings, :missing_readings, :missing_mpan_mprn, :missing_reading_date, :invalid_reading_date]
```
Comment (jamesalmond) [inline app/models/amr_reading_warning.rb]: remove blank readings unless required for empty strings

File: app/services/amr/data_feed_validator.rb [hunk 10]
```diff
@@ -7,13 +7,21 @@ def initialize(config, array_of_rows)
...
+      handle_header
       @array_of_rows.delete_if { |row| invalid_row?(row) }
       @array_of_rows
     end
 
   private
 
+    def handle_header
+      if @array_of_rows.first.join(',') == @config.header_example
+        @array_of_rows.shift
```
Comment (jamesalmond) [inline app/services/amr/data_feed_validator.rb]: Use slice. Don't mutate @array_of_rows in this class.

File: app/services/amr/data_feed_validator.rb [hunk 11]
```diff
@@ -7,13 +7,21 @@ def initialize(config, array_of_rows)
 
     def perform
       @array_of_rows = sort_out_off_by_one_array(@array_of_rows) if @config.handle_off_by_one
-      @array_of_rows.shift(@config.number_of_header_rows)
+      handle_header
       @array_of_rows.delete_if { |row| invalid_row?(row) }
```
Comment (jamesalmond) [inline app/services/amr/data_feed_validator.rb]: Reject

File: app/services/amr/process_amr_reading_data.rb [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+      inserted_record_count = AmrDataFeedReading.count - records_before
+
+      amr_data_feed_import_log.update(records_imported: inserted_record_count, records_upserted: upserted_record_count)
+    end
+
+    def create_warnings
+      updated_warnings = @amr_reading_data.warnings.map do |warning|
+        {
+          amr_data_feed_import_log_id: @amr_data_feed_import_log.id,
+          warning_message: AmrReadingData::WARNINGS[warning[:warning]],
```
Comment (jamesalmond) [inline app/services/amr/process_amr_reading_data.rb]: Don't persist messages

File: app/views/admin/amr_uploaded_readings/_first_ten_reading_rows.html.erb [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+    <thead>
+      <tr>
+        <th>Meter ID</th>
+        <th>Mpan Mprn</th>
+        <th>Reading Date</th>
+        <th colspan="48">Readings</th>
+      </tr>
+    </thead>
+    <tbody>
+      <% valid_reading_data.first(10).each do |row| %>
```
Comment (jamesalmond) [inline app/views/admin/amr_uploaded_readings/_first_ten_reading_rows.html.erb]: first(10) moves to controller

File: app/views/admin/amr_uploaded_readings/_first_ten_reading_rows.html.erb [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+        <tr>
+          <td><%= row["meter_id"] %></td>
+          <td><%= row["mpan_mprn"] %></td>
+          <td><%= nice_dates Date.parse(row["reading_date"]) %></td>
```
Comment (jamesalmond) [inline app/views/admin/amr_uploaded_readings/_first_ten_reading_rows.html.erb]: Show raw reading date

File: spec/models/amr_reading_data_spec.rb [hunk 15]
```diff
@@ -12,81 +13,167 @@
...
+        expect(amr_reading.warnings.count).to be 1
+        expect(amr_reading.warnings.first[:warning]).to eq :invalid_reading_date
       end
     end
   end
 end
+
+
+
+  # describe 'handles when a row of readings is missing' do
```
Comment (jamesalmond) [inline spec/models/amr_reading_data_spec.rb]: Delete these

File: spec/services/amr/csv_parser_and_upserter_warnings_spec.rb [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+      importer.perform
+      importer.inserted_record_count
+    end
+
+    context 'banes' do
+      before(:each) do
+        FileUtils.mkdir_p config.local_bucket_path
+      end
+
+      it 'should parse a simple file' do
```
Comment (jamesalmond) [inline spec/services/amr/csv_parser_and_upserter_warnings_spec.rb]: Review whether necessary

File: app/models/amr_reading_warning.rb [hunk 17]
```diff
@@ -24,5 +25,21 @@
...
+    1 => :missing_readings,
+    2 => :missing_mpan_mprn,
+    3 => :missing_reading_date,
+    4 => :invalid_reading_date
+  }.freeze
+
   enum warning: [:blank_readings, :missing_readings, :missing_mpan_mprn, :missing_reading_date, :invalid_reading_date]
+
+  def messages
+    warning_symbols.map { |warning_symbol| AmrReadingData::WARNINGS[warning_symbol] }.join(', ') + 'HELLO'
```
Comment (jamesalmond) [inline app/models/amr_reading_warning.rb]: HELLO? HELLO?!

File: app/models/amr_reading_data.rb [hunk 18]
```diff
@@ -61,26 +61,29 @@ def any_valid_readings?
...
+      warnings << :missing_mpan_mprn if reading[:mpan_mprn].blank?
+      warnings << :missing_reading_date if reading_date.blank?
+      warnings << :invalid_reading_date unless reading_date.present? && valid_reading_date?(reading_date)
+
+      reading[:warnings] = warnings if warnings.any?
     end
   end
 
   def missing_readings?(readings)
     readings.compact.size < (48 - @missing_reading_threshold)
```
Comment (jamesalmond) [inline app/models/amr_reading_data.rb]: Do we need this if we have blanks?

**Pull request merged**